src/data/monoid_algebra.lean

```diff
@@ -159,10 +209,59 @@ finsupp.semimodule G k
 instance [ring k] : module k (monoid_algebra k G) :=
 finsupp.module G k
 
-lemma single_mul_single [semiring k] [monoid G] {a₁ a₂ : G} {b₁ b₂ : k} :
-  (single a₁ b₁ : monoid_algebra k G) * single a₂ b₂ = single (a₁ * a₂) (b₁ * b₂) :=
-(sum_single_index (by simp only [_root_.zero_mul, single_zero, sum_zero])).trans
-  (sum_single_index (by rw [_root_.mul_zero, single_zero]))
+instance [comm_semiring k] [monoid G] : algebra k (monoid_algebra k G) :=
+{ to_fun := single 1,
+  map_one' := rfl,
+  map_mul' := λ x y, by rw [single_mul_single, one_mul],
+  map_zero' := single_zero,
+  map_add' := λ x y, single_add,
+  smul_def' := λ r a, by { ext x, exact smul_apply.trans (single_one_mul_apply _ _ _).symm },
+  commutes' := λ r f, show single 1 r * f = f * single 1 r,
+    by ext; rw [single_one_mul_apply, mul_single_one_apply, mul_comm] }
+
+@[simp] lemma coe_algebra_map [comm_semiring k] [monoid G] :
+  (algebra_map k (monoid_algebra k G) : k → monoid_algebra k G) = single 1 :=
+rfl
+
+/-- Any monoid homomorphism `G →* R` can be lifted to an algebra homomorphism
+`monoid_algebra k G →ₐ[k] R`. -/
+def lift [comm_semiring k] [monoid G] {R : Type u₃} [semiring R] [algebra k R] :
+  (G →* R) ≃ (monoid_algebra k G →ₐ[k] R) :=
+{ inv_fun := λ f, ((f : monoid_algebra k G →+* R) : monoid_algebra k G →* R).comp (of k G),
+  to_fun := λ F, { to_fun := λ f, f.sum (λ a b, b • F a),
+    map_one' := by { rw [one_def, sum_single_index, one_smul, F.map_one], apply zero_smul },
+    map_mul' :=
+      begin
+        intros f g,
+        rw [mul_def, finsupp.sum_mul, finsupp.sum_sum_index];
+          try { intros, simp only [zero_smul, add_smul], done },
+        refine finset.sum_congr rfl (λ a ha, _), simp only [],
+        rw [finsupp.mul_sum, finsupp.sum_sum_index];
+          try { intros, simp only [zero_smul, add_smul], done },
+        refine finset.sum_congr rfl (λ a' ha', _), simp only [],
+        rw [sum_single_index, F.map_mul, algebra.mul_smul_comm, algebra.smul_mul_assoc,
+          smul_smul, mul_comm],
+        apply zero_smul
+      end,
+    map_zero' := sum_zero_index,
+    map_add' := λ f g, by rw [sum_add_index]; intros; simp only [zero_smul, add_smul],
+    commutes' := λ r, by rw [coe_algebra_map, sum_single_index, F.map_one, algebra.smul_def,
+      mul_one]; apply zero_smul },
+  left_inv := λ f,
+    begin
+      ext x, dsimp,
+      -- These `norm_cast`s used to not be necessary,
+      -- and I'm afraid I've broken something when merging with master.
+      norm_cast, dsimp,
```

Comment:
@gebner, @pnmadelaine, after joining with `master` I've had to insert a `norm_cast` here that previously wasn't required. I'm worried I've messed something up with `norm_cast` attributes. Is there any chance one of you could take a look at this?

Reply:
Oh wow, this is a fun goal: `⇑↑↑{...} (single x 1) = ⇑f x`.

My first guess is that we need a simp lemma that rewrites single-double-arrow-double-single-arrow into double-arrow.  If you turn on the following debugging options, then you can see what lemmas norm_cast applies:
```lean
set_option trace.norm_cast true
set_option trace.simplify.rewrite true
```

I've added `@[simp]` to `coe_monoid_hom'` and `coe_add_monoid_hom'`.  I am not sure why this used to work before.  I've also fixed the non-terminal (d)simps here.

Reply:
Great, thanks! I've copied your fix over to `add_monoid_algebra` as well.